Sofa/Component/Mapping/Linear/src/sofa/component/mapping/linear/DistanceToPlaneMapping.inl

```diff
@@ -0,0 +1,213 @@
+    if (normalNorm<std::numeric_limits<double>::epsilon())
+    {
+        msg_error(this)<<" planeNormal data has null norm.";
+        this->d_componentState.setValue(core::objectmodel::ComponentState::Invalid);
+        return;
+    }
+    d_planeNormal.setValue(d_planeNormal.getValue()/normalNorm);
+
+    constexpr Size inCoordSize = typename Coord_t<TIn>::Size();
+    constexpr Size inDerivSize = typename Deriv_t<TIn>::Size();
+    Size inSize = this->getFromModel()->getSize();
+    this->getToModel()->resize( inSize );
+
+    const auto planeNormal = d_planeNormal.getValue();
+
+    J.compressedMatrix.resize( inSize, inSize*inCoordSize );
+
+    for (Size i = 0; i < inSize; i++)
+    {
+        const size_t col = i * inCoordSize;
+        J.compressedMatrix.startVec(i);
+
+        for (Size j = 0; j < inDerivSize; j++ )
+        {
+            if constexpr (sofa::type::isRigidType<TIn>)
+            {
+                if (i>=Deriv_t<TIn>::Pos::size())
+                    break;
+            }
+            J.compressedMatrix.insertBack( i, col ) = planeNormal[i];
+        }
+        J.compressedMatrix.finalize();
+
+    }
+
+    this->d_componentState.setValue(core::objectmodel::ComponentState::Valid);
+}
+
+template <class TIn>
+void DistanceToPlaneMapping<TIn>::apply(const core::MechanicalParams *mparams, Data<VecCoord_t<TOut>>& out, const Data<VecCoord_t<TIn>>& in)
+{
+    if (this-> d_componentState.getValue() != sofa::core::objectmodel::ComponentState::Valid)
+        return;
+
+    auto writeOut = helper::getWriteAccessor(out);
+    const auto readIn = helper::getReadAccessor(in);
+
+    const auto planeNormal = d_planeNormal.getValue();
+
+    double planeDistanceToOrigin;
+    if constexpr (!type::isRigidType<TIn>)
+        planeDistanceToOrigin = dot(d_planePoint.getValue(),planeNormal);
+    else
+        planeDistanceToOrigin  = dot(d_planePoint.getValue().getCenter(),planeNormal.getVCenter());
+
+
+    for ( unsigned i = 0; i<readIn.size(); i++ )
+    {
+        if constexpr (!type::isRigidType<TIn>)
+            writeOut[i] = type::dot(readIn[i],planeNormal) - planeDistanceToOrigin;
+        else
+            writeOut[i] = type::dot(readIn[i].getCenter(),planeNormal.getVCenter()) - planeDistanceToOrigin;
+    }
+}
+
+template <class TIn>
+void DistanceToPlaneMapping<TIn>::applyJ(const core::MechanicalParams *mparams, Data<VecDeriv_t<TOut>>& out, const Data<VecDeriv_t<TIn>>& in)
+{
+    if (this-> d_componentState.getValue() != sofa::core::objectmodel::ComponentState::Valid)
+        return;
+
+    auto writeOut = helper::getWriteAccessor(out);
+    const auto readIn = helper::getReadAccessor(in);
+    const auto planeNormal = d_planeNormal.getValue();
+
+    for ( unsigned i = 0; i<readIn.size(); i++ )
+    {
+        if constexpr (!type::isRigidType<TIn>)
+            writeOut[i] = type::dot(readIn[i],planeNormal);
+        else
+            writeOut[i] = type::dot(readIn[i].getVCenter(),planeNormal.getVCenter());
+    }
```

Comment:
The use of getCPos (DataTypes::getCPos) should allow the removal of the if constexpr.